packages/react-dom/src/server/ReactDOMFloatServer.js

```diff
@@ -0,0 +1,587 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import ReactDOMSharedInternals from 'shared/ReactDOMSharedInternals';
+import {
+  validatePreloadResourceDifference,
+  validateStyleResourceDifference,
+  validateStyleAndHintProps,
+  validateLinkPropsForStyleResource,
+  validateLinkPropsForPreloadResource,
+  validatePreloadArguments,
+  validatePreinitArguments,
+} from '../shared/ReactDOMResourceValidation';
+
+type Props = {[string]: mixed};
+
+type ResourceType = 'style' | 'font';
+
+type PreloadProps = {
+  rel: 'preload',
+  as: ResourceType,
+  href: string,
+  [string]: mixed,
+};
+type PreloadResource = {
+  type: 'preload',
+  as: ResourceType,
+  href: string,
+  props: PreloadProps,
+  flushed: boolean,
+};
+
+type StyleProps = {
+  rel: 'stylesheet',
+  href: string,
+  'data-rprec': string,
+  [string]: mixed,
+};
+type StyleResource = {
+  type: 'style',
+  href: string,
+  precedence: string,
+  props: StyleProps,
+
+  flushed: boolean,
+  inShell: boolean, // flushedInShell
+  hint: PreloadResource,
+};
+
+export type Resource = PreloadResource | StyleResource;
+
+export type Resources = {
+  // Request local cache
+  preloadsMap: Map<string, PreloadResource>,
+  stylesMap: Map<string, StyleResource>,
+
+  // Flushing queues for Resource dependencies
+  explicitPreloads: Set<PreloadResource>,
+  implicitPreloads: Set<PreloadResource>,
+  precedences: Map<string, Set<StyleResource>>,
+
+  // Module-global-like reference for current boundary resources
+  boundaryResources: ?BoundaryResources,
+};
+
+export function createResources(): Resources {
+  return {
+    // persistent
+    preloadsMap: new Map(), // preloadResources
+    stylesMap: new Map(), // styleResources
+
+    // cleared on flush
+    explicitPreloads: new Set(), // explicitPreloads
+    implicitPreloads: new Set(), // add bootstrap script to implicit preloads
+    precedences: new Map(),
+
+    // like a module global for currently rendering boundary
+    boundaryResources: null,
+  };
+}
+
+export type BoundaryResources = Set<StyleResource>;
+
+export function createBoundaryResources(): BoundaryResources {
+  return new Set();
+}
+
+export function mergeBoundaryResources(
+  target: BoundaryResources,
+  source: BoundaryResources,
+) {
+  const iter = source.values();
+  let {value, done} = iter.next();
```

Comment:
The fastest and most compatible way to iterate through Sets and Maps is `forEach`. That's what we use elsewhere. It needs a closure but at least it's only one instead of per iteration.

Reply:
TIL